processing/src/test/java/io/druid/segment/incremental/IncrementalIndexStorageAdapterTest.java

```diff
@@ -458,4 +458,157 @@ public Object apply(Cursor cursor)
+                          ImmutableMap.<String, Object>of("billy", "v1")
+                      )
+                  );
+                }
+                catch (Exception ex) {
+                  throw new RuntimeException(ex);
+                }
+
+                DimensionSelector dimSelectorB = cursor.makeDimensionSelector(
+                    new DefaultDimensionSpec(
+                        "billy",
+                        "billy"
+                    )
+                );
+                //index gets more rows at this point, while other thread is iterating over the cursor
+                try {
+                  index.add(
+                      new MapBasedInputRow(
+                          timestamp,
+                          Lists.newArrayList("billy"),
+                          ImmutableMap.<String, Object>of("billy", "v2")
+                      )
+                  );
+                  index.add(
+                      new MapBasedInputRow(
+                          timestamp,
+                          Lists.newArrayList("billy2"),
+                          ImmutableMap.<String, Object>of("billy2", "v3")
+                      )
+                  );
+                }
+                catch (Exception ex) {
+                  throw new RuntimeException(ex);
+                }
+
+                DimensionSelector dimSelectorC = cursor.makeDimensionSelector(
+                    new DefaultDimensionSpec(
+                        "billy",
+                        "billy"
+                    )
+                );
+
+                DimensionSelector dimSelectorD = cursor.makeDimensionSelector(
+                    new DefaultDimensionSpec(
+                        "billy2",
+                        "billy2"
+                    )
+                );
+                //index gets more rows at this point, while other thread is iterating over the cursor
+                try {
+                  index.add(
+                      new MapBasedInputRow(
+                          timestamp,
+                          Lists.newArrayList("billy"),
+                          ImmutableMap.<String, Object>of("billy", "v3")
+                      )
+                  );
+                  index.add(
+                      new MapBasedInputRow(
+                          timestamp,
+                          Lists.newArrayList("billy3"),
+                          ImmutableMap.<String, Object>of("billy3", "")
+                      )
+                  );
+                }
+                catch (Exception ex) {
+                  throw new RuntimeException(ex);
+                }
+
+                DimensionSelector dimSelectorE = cursor.makeDimensionSelector(
+                    new DefaultDimensionSpec(
+                        "billy3",
+                        "billy3"
+                    )
+                );
+
+                // and then, cursoring continues in the other thread
+                while (!cursor.isDone()) {
```

Comment:
because cant't make dimensionSelectorA from empty cursor from empty index, so I just starting from a non-empty index. Then `cursor.isDone()` is not always true